### ISSUE TITLE
TypeError: parent.addNewSubsegmentWithoutSampling is not a function

### ISSUE THREAD

**Issue by yvele** [aws/aws-xray-sdk-node#544]
When updating dependencies:
- aws-sdk: 2.1231.0 → **2.1254.0**
- aws-xray-sdk-core: 3.3.8 → **3.4.0**

I got a Node.js error

> TypeError: parent.addNewSubsegmentWithoutSampling is not a function

With the following callstack:

```none
 TypeError: parent.addNewSubsegmentWithoutSampling is not a function
     at features.constructor.captureAWSRequest [as customRequestHandler] (webpack://my-package/./node_modules/aws-xray-sdk-core/dist/lib/patchers/aws_p.js?:67:29)
     at features.constructor.addAllRequestListeners (webpack://my-package/./node_modules/aws-sdk/lib/service.js?:302:12)
     at features.constructor.makeRequest (webpack://my-package/./node_modules/aws-sdk/lib/service.js?:222:10)
     at features.constructor.svc.<computed> [as executeStatement] (webpack://my-package/./node_modules/aws-sdk/lib/service.js?:706:23)
     at DataService.query (webpack://my-package/./node_modules/@hapticmedia/aws-rds/lib/DataService/DataService.js?:99:37)
     at getShowcaseReleases (webpack://my-package/./src/resolvers/getShowcaseReleases.js?:45:16)
     at ResolverRouter.resolveEventBatch (webpack://my-package/./node_modules/my-helper/lib/ResolverRouter/ResolverRouter.js?:270:33)
     at ResolverRouter.resolveEvent (webpack://my-package/./node_modules/my-helper/lib/ResolverRouter/ResolverRouter.js?:323:12)
     at handleCore (webpack://my-package/./src/main.js?:100:126)
     at Runtime.handle [as handler] (webpack://my-package/./src/main.js?:106:22)
```

**Comment by carolabadeer** [comment]
Can you please provide a code snippet of the segments / subsegments you are creating in this instance that lead you to this error? Does this only happen when you make calls to the aws sdk (v2) or do you also see it with aws sdk v3 and HTTP calls?

**Comment by willarmiros** [comment]
Hi folks - are you all running in Lambda?

**Comment by carolabadeer** [comment]
Hi all. I have been trying to reproduce the error you mentioned above through testing using [our Node SDK sample app ](https://github.com/aws-samples/aws-xray-sdk-node-sample) which uses the latest version of the node SDK (3.4.0) and AWS SDK v2 as well as through lambda but still cannot see the error you are reporting. Can you please provide more context on your applications and, if possible, a more complete code example that we can use to reproduce this issue?

Here is what I have tried so far using the latest version. I do not run into any errors and can see a complete service map on the AWS X-Ray console (including the instrumented AWS SDK calls). 
- Running the [X-Ray Node SDK express sample app](https://github.com/aws-samples/aws-xray-sdk-node-sample) using the instructions in the corresponding repo's ReadMe. Both endpoints for the AWS SDK call and HTTP call are working and sending traces to the X-Ray backend on my end. 
- Using the X-Ray SDK locally to write a function that makes + instruments AWS SDK calls to S3. This is similar to the following point but tests the SDK outside of a lambda environment. 
- Creating a lambda function locally, installing the X-Ray SDK and AWS SDK v2 using `npm install` commands, running the `zip -r lambdaFunc.zip . ` command to create a zip containing the lambda function and `node_modules` folder, and uploading that .zip file to lambda. This is the lambda function:
```
const { Segment } = require('aws-xray-sdk');
const xray = require('aws-xray-sdk');

const AWS = xray.captureAWS(require('aws-sdk'));

exports.handler = async (event, context) => {        
    let toSample = false;  // this can be modified - set to false to skip to the case of an unsampled facade segment and execute the line that seems to be throwing an error

    if(toSample){
        let facade = xray.getSegment();
        let test = facade.addNewSubsegment('sqs-subsegment-sampled');
        console.log("doing batch work - sampled");
          try {
            const s3 = new AWS.S3();
            let data = await s3.listBuckets().promise();
            console.log(data);
          } catch (error) {
            console.log("retrieveFromS3 error:", error);
            throw error;
          }
        test.close()
    }
    else {
        let facade = xray.getSegment();
        facade.notTraced = true
        let unsampled = facade.addNewSubsegmentWithoutSampling('sqs-subsegment-unsampled')
          try {
            const s3 = new AWS.S3();
            let data = await s3.listBuckets().promise();
            console.log(data);
          } catch (error) {
            console.log("retrieveFromS3 error:", error);
            throw error;
          }
        unsampled.close();
    }
    return 'Success';
}
```

If possible, please try to test one of the options above and let us know if you still see the same error on your end. Otherwise, please provide a more complete code snippet (like a complete lambda function) that we can use to reproduce the issue.

**Comment by cs17** [comment]
> Hi folks - are you all running in Lambda?

Yes, we are running in Lambda with Nodejs16. 



> Hi all. I have been trying to reproduce the error mentioned above through testing using [our Node SDK sample app](https://github.com/aws-samples/aws-xray-sdk-node-sample), which uses the latest version of the node SDK (3.4.0) and AWS SDK v2, as well as through lambda but still cannot see the error you are reporting. Can you please provide more context on your applications and, if possible, a more complete code example that we can use to reproduce this issue?
> 
> Here is what I have tried so far using the latest version. I do not run into any errors and can see a complete service map on the AWS X-Ray console (including the instrumented AWS SDK calls).
> 
> * Running the [X-Ray Node SDK express sample app](https://github.com/aws-samples/aws-xray-sdk-node-sample) using the instructions in the corresponding repo's ReadMe. Both endpoints for the AWS SDK call and HTTP call are working and sending traces to the X-Ray backend on my end.
> * Using the X-Ray SDK locally to write a function that makes + instruments AWS SDK calls to S3. This is similar to the following point but tests the SDK outside of a lambda environment.
> * Creating a lambda function locally, installing the X-Ray SDK and AWS SDK v2 using `npm install` commands, running the `zip -r lambdaFunc.zip .` command to create a zip containing the lambda function and `node_modules` folder, and uploading that .zip file to lambda. This is the lambda function:
> 
> ```
> const { Segment } = require('aws-xray-sdk');
> const xray = require('aws-xray-sdk');
> 
> const AWS = xray.captureAWS(require('aws-sdk'));
> 
> exports.handler = async (event, context) => {        
>     let toSample = false;  // this can be modified - set to false to skip to the case of an unsampled facade segment and execute the line that seems to be throwing an error
> 
>     if(toSample){
>         let facade = xray.getSegment();
>         let test = facade.addNewSubsegment('sqs-subsegment-sampled');
>         console.log("doing batch work - sampled");
>           try {
>             const s3 = new AWS.S3();
>             let data = await s3.listBuckets().promise();
>             console.log(data);
>           } catch (error) {
>             console.log("retrieveFromS3 error:", error);
>             throw error;
>           }
>         test.close()
>     }
>     else {
>         let facade = xray.getSegment();
>         facade.notTraced = true
>         let unsampled = facade.addNewSubsegmentWithoutSampling('sqs-subsegment-unsampled')
>           try {
>             const s3 = new AWS.S3();
>             let data = await s3.listBuckets().promise();
>             console.log(data);
>           } catch (error) {
>             console.log("retrieveFromS3 error:", error);
>             throw error;
>           }
>         unsampled.close();
>     }
>     return 'Success';
> }
> ```
> 
> If possible, please try to test one of the options above and let us know if you still see the same error on your end. Otherwise, please provide a more complete code snippet (like a complete lambda function) that we can use to reproduce the issue.

We managed to simulate the issue on our side. 

Our lambda function uses one shared layer, which still uses aws-xray-sdk-core `v3.3.8`, as the shared layer was deployed a few weeks ago. (Before the release of `v3.4.0`)

For our application lambda, the codebuild installs the latest `v3.4.0` for our function, as we are using `"aws-xray-sdk-core": "^3.3.4"`. (Compatible with version)

Example code with issue:
```
const helperLayer = require("helperLayer"); // Shared layer - using v3.3.8
const AWSXRay = require('aws-xray-sdk-core'); // This lambda - using v3.4.0
const AWS = AWSXRay.captureAWS(require('aws-sdk'));
...
```

We solved the issue by redeploying the shared layer to allow it to pull the latest version of aws-xray-sdk-core, so our shared layer and application lambda is using the same & latest version of aws-xray-sdk-core. 


Another trick to overcome this is to do this, not sure why. :no_mouth:
```
const AWSXRay = require('aws-xray-sdk-core'); // This lambda - using v3.4.0
const AWS = AWSXRay.captureAWS(require('aws-sdk'));
const helperLayer = require("helperLayer"); // Shared layer - using v3.3.8
...
```

**Comment by yvele** [comment]
> Hi folks - are you all running in Lambda?

Yes Lambda with Node.js v16

**Comment by yvele** [comment]
@carolabadeer the issue has been fixed? 🤔 There is no new release after the 3.4.0 https://github.com/aws/aws-xray-sdk-node/releases could you provide us with more information please? Thanks
